HIP/hip-859.md

```diff
@@ -0,0 +1,158 @@
+
+That means if the EVM consumes for example 55_500 gas, but the user is passing 400_000 gas limit, they will be charged the `max(55_500, 320_000)`, so it would be 320_000. That is the value calculated after we return the 20% of the gasLimit back to the user, which is 80_000.
+The logic compares the bigger of the two values - the gas consumed by the EVM and the 80% of the `gasLimit` that the user has passed.
+
+In this way we lose track of the actual gas that was consumed by the EVM.
+
+This HIP will add a new field `gas_consumed` to the `contract_result` table in the mirror node, which will contain the actually consumed gas by the EVM.
+
+```
+create table if not exists contract_result
+(
+...
+gas_limit            bigint       not null,
+gas_consumed         bigint       null,
+gas_used             bigint       null,
+...
+)
+```
+
+## Motivation
+
+Adding a `gas_consumed` field in the `contract_result` table would enrich user experience by providing more information about the transaction. 
+This would allow users to better understand the gas refund policy and the actual gas consumption of their transactions. This would also allow users to understand how heavy their executed transaction was for the EVM.
+
+## Rationale
+
+Consensus nodes export ContractActions sidecar records for each transaction after the introduction of [HIP-513](./hip-513.md). Each sidecar record contains a `gasUsed` field for the given action that the record represents (different type of call or contract creation).
+This value directly represents the gas consumed by the EVM, without taking into account the gas refund policy. This makes it possible to calculate the overall gas consumption for the transaction, during the ContractActions sidecar record importing.  
+
+If we sum up all of the `gasUsed` values of each ContractActions sidecar for a given transaction on top of the initial intrinsic gas, which is the initial gas, charged prior to the transaction execution, then we would get accurate value for the `gasConsumed` field.
+
+## User stories
+
+1. As a user, I want to see how much gas was consumed by the EVM during the execution of my transaction, when I retrieve a transaction info from the mirror-node REST APIs, so that I know if I can optimise my transaction cost
+2. As a user, I want to see and compare how much gas I was charged for and how much gas was consumed by the EVM, when I inspect my transaction in hashscan
+
+## Specification
+
+In order to calculate the `gasUsed` field, the mirror node should have `CONTRACT_ACTION` and `CONTRACT_BYTECODE` sidecars enabled. Then, we would use an enhancement logic in the `mirror-node-importer`,
+which will iterate over all `ContractAction` sidecar records for the transaction we want to import and calculate the `gasConsumed` field for.
+
+The new logic would have a mechanism to first calculate the intrinsic gas cost for the transaction being imported. We can inspect if the transaction has a `ContractByteCode` sidecar
```

Comment:
This logic would not work for contract creates pre-sidecar. New mirror nodes could start and process all historical data so we have to handle this scenario. Same problem if sidecars are disabled. For those should we set gasConsumed null, zero, or equal to gasUsed?

Reply:
Equal to gasUsed is probably best as null and 0 would just mean the caller would need to use gasUsed

Reply:
I updated the HIP with a description about setting the field as null. I think having `gasConsumed` equal to `gasUsed` in those cases, could be confusing, since we are telling the users that the EVM consumed as much gas as it was charged using the refund policy. This is often not the case and this case it would be a guess and we would mislead the user with returning wrong value for this field. If we have `null`, this would mean the value is absent and it's up to the users to decide how to use the gas information, if they need to handle it.